### PR TITLE
Enforce valid function names in the packages/block-library/src/*/*.php files

### LINKED ISSUE #52769: Enforce valid function names in `packages/block-library/*/src/index.php` files
## What problem does this address?
Follow-up from https://github.com/WordPress/gutenberg/pull/52696.
Functions with `gutenberg_` prefixes should not be permitted in the `block-library` package.

Please see:

1. https://github.com/WordPress/gutenberg/pull/51978
2. https://github.com/WordPress/gutenberg/pull/52696#pullrequestreview-1536235865
3. https://github.com/WordPress/gutenberg/issues/44151#issuecomment-1613503035

Props:  @tellthemachines, @ockham

## What is your proposed solution?
After the merge of https://github.com/WordPress/gutenberg/pull/52696, a new phpcs sniff should be implemented to check that these functions do not exist in the `block-library` package.

### PR DESCRIPTION
**Pull request by anton-vlasenko**
## What?
This PR is focused on implementing a new PHPCS sniff, namely `Gutenberg.NamingConventions.ValidBlockLibraryFunctionName`.
Fixes https://github.com/WordPress/gutenberg/issues/52769.

## Why?
All the PHP function names within the `./packages/block-library/src/*/*.php` files should start with one of the following prefixes:

- `block_core_<folder_name>`
- `render_block_core_<folder_name>`
- `register_block_core_<folder_name>`

Here, `<folder_name>` represents the name of the folder where the associated file is situated. The folder name is converted to lowercase, and any characters other than letters and digits are substituted with underscores.

The introduction of this new PHPCS sniff is needed to ensure strict adherence to this naming pattern.

## How?
The proposed `ValidBlockLibraryFunctionNameSniff` is configured to examine PHP function names contained within the `./packages/block-library/src/*/*.php` files.

Whenever a function name deviates from the naming convention listed above, the PHPCS sniff will add an error to the PHPCS report.

## Testing Instructions
1. Make sure that CI jobs pass.
2. Open the `packages/block-library/src/calendar/index.php` file in your preferred code editor.
3. Add a new PHP function to the file:
```php
/**
 * A test function.
 */
function block_core_calendar_() {

}
```
4. Run the linter (`npm run lint:php` or `composer run lint`, based on your environment setup).
5. Notice the linter error displayed in the console:
```
The function name "block_core_calendar_()" is invalid because PHP function names in this file should start with one of the following prefixes: "block_core_calendar", "render_block_core_calendar", "register_block_core_calendar".
```
6. Rename the php function to `block_core_clndr`.
7. Rerun the linter.
8. Notice the linter error displayed in the console:
```
The function name "block_core_clndr()" is invalid because PHP function names in this file should start with one of the following prefixes: "block_core_calendar", "render_block_core_calendar", "register_block_core_calendar".
```
9. Rename the php function to `block_core_calendar`.
10. Rerun the linter.
11. Confirm the absence of linter errors in the console.
12. [Optional] Perform testing with other allowed prefixes, like `render_block_core_calendar` and `register_block_core_calendar`.

### CODE REVIEW

**Comment by gziolo** [comment]
See my comment in the parent issue: https://github.com/WordPress/gutenberg/issues/52769#issuecomment-1670864036. I think it's fine to add code comments that ignore existing violations and enforce good practices moving on.

File: phpcs.xml.dist
Comment (gziolo) [inline phpcs.xml.dist]: At the moment, we have only `index.php` files, but maybe we should make it less restrictive:
```suggestion
		<include-pattern>./packages/block-library/src/*/*.php</include-pattern>
```

I don't know if that will be useful so maybe that change can be applied if it's as simple as updating this line.

Comment (anton-vlasenko) [reply phpcs.xml.dist]: Yes, this should work. Thanks for the suggestion.

**Comment by gziolo** [comment]
It seems to be in a good shape, let me know when I can give a final review and approve.

**Comment by anton-vlasenko** [comment]
@gziolo

> It seems to be in a good shape, let me know when I can give a final review and approve.

I had to address an edge case and write proper testing instructions.
Yes, please give it a final review. 🙏 Thank you!

File: packages/block-library/src/avatar/index.php
Comment (gziolo) [inline packages/block-library/src/avatar/index.php]: This and other files are going to be copied to WordPress core. Should this rule be present in both places and get prefixed with `WordPress.`? Maybe it's fine to have it only in Gutenberg, just flagging the fact that these files are tricky to maintain.

File: packages/block-library/src/navigation/index.php
Comment (gziolo) [inline packages/block-library/src/navigation/index.php]: I'm a little bit confused here with the `gutenberg_` prefix and the usage of `gutenberg_should_block_use_interactivity_api`. The latter will trigger a fatal error in WordPress core when the files gets copied. This code should leave outside of this file, so in that regard the lining rule confirms that it's very useful.

Aside, we should also enforce that you can't use functions prefixed with `gutenberg_` with core blocks in the same set of files as they don't exist in WP core.

As for this PR, it's fine to proceed with the ignore comment, but the usage of `gutenberg_should_block_use_interactivity_api` needs to be replaced with code present in WP core before we start backporting changes to WP core. @luisherranz, are you aware of this issue?

File: phpcs.xml.dist
Comment (gziolo) [inline phpcs.xml.dist:138]: I see functions that start `build_template_part_`, so maybe we could open also `build_block_core_template_part` in case folks want to use that. I don't have strong opinions on that, as they could also do `block_core_tempalte_part_build`.

File: packages/block-library/src/footnotes/index.php
Comment (gziolo) [inline packages/block-library/src/footnotes/index.php]: Unrelated to this PR. I'm slightly surprised that the Footnotes block works correctly with WordPress 6.3. We don't guard calling these WP hooks when they are present in the WP core, so essential we register them twice with the Gutenberg plugin. We should double-check it and if necessary move all the code that handles WP hooks to another file in `lib/compat/` with some sort of check if it was already applied and override them only if necessary. At the same time, we would have to move the same code to a different place in WordPress core. @ellatrix, what do you think?

File: packages/block-library/src/navigation/index.php
Comment (anton-vlasenko) [reply packages/block-library/src/navigation/index.php]: > Aside, we should also enforce that you can't use functions prefixed with gutenberg_ with core blocks in the same set of files as they don't exist in WP core.

Sure. This can enforced. I will update the PR, @gziolo.

Comment (luisherranz) [reply packages/block-library/src/navigation/index.php]: > @luisherranz, are you aware of this issue?

Not at all. I don't have any experience backporting stuff to WP core yet, nor with how backporting requirements affect the code of Gutenberg. But happy to fix things if you give me instructions on what needs to be done, tho.

cc: @westonruter (pinging because Weston introduced the `gutenberg_should_block_use_interactivity_api` function, just in case)

File: packages/block-library/src/avatar/index.php
Comment (anton-vlasenko) [reply packages/block-library/src/avatar/index.php]: @gziolo To use the `WordPress.` in the rule's name, it would need to become a part of the [WordPress-Coding-Standards](https://github.com/WordPress/WordPress-Coding-Standards) package (`wpcs`). However, getting this rule into the `wpcs` package might be challenging.

`wpcs` is currently undergoing a major overhaul. It's possible that you may need to wait for an undefined period of time until it's released.

Since this rule is currently essential for Gutenberg, my suggestion is to commit it to Gutenberg first. If the `wpcs` contributors agree that it should be included in `wpcs`, it can always be removed from the Gutenberg package.

To address the concern about maintainability, I will relocate the `//phpcs:disable` and `//phpcs:enable` statements from the doc blocks to prevent any interference.

File: packages/block-library/src/navigation/index.php
Comment (gziolo) [reply packages/block-library/src/navigation/index.php]: Related to that, I think it could be valuable to include a mandatory banner in every PHP file we know gets backported to WordPress core with the link to the document explaining all the expected requirements enforced by the complex process. It will also give a hint to folks whenever they try to update those files directly in the WordPress core which is going to be replaced the next time someone runs `npm run build` ...

Obviously not in this PR.

File: phpcs.xml.dist
Comment (anton-vlasenko) [reply phpcs.xml.dist:138]: I don't have a preference either. However, please note that the linting rule not only checks the prefix name but also the directory name, as requested [here](https://github.com/WordPress/gutenberg/issues/52769#issuecomment-1669540435).

Comment (gziolo) [reply phpcs.xml.dist:138]: It's fine as is for now, but we should keep monitoring and extending the list if that is useful.

File: packages/block-library/src/avatar/index.php
Comment (gziolo) [reply packages/block-library/src/avatar/index.php]: We might have linting disabled on files copied from Gutenberg through WP packages, so you can double check that first before applying any changes. I guess we can live with the `Gutenberg.` specific rule if it doesn't throw errors on CI in WP core.

Comment (anton-vlasenko) [reply packages/block-library/src/avatar/index.php]: @gziolo 
The `//phpcs:disable Gutenberg.NamingConventions.ValidBlockLibraryFunctionName` statements shouldn't generate any linter errors in Core. However, personally, I'm not fond of the idea of having these comments in Core code.
I think I will add the possibility to ignore specific functions via the linter rule's configuration in phpcs.xml.dist. This way, we can avoid cluttering Core code with these phpcs: statements. I'll make the necessary updates to the PR and let you know.

Comment (gziolo) [reply packages/block-library/src/avatar/index.php]: Yes, we could list these function names as exceptions in the config file. Not ideal, but it should work.

File: packages/block-library/src/navigation/index.php
Comment (anton-vlasenko) [reply packages/block-library/src/navigation/index.php]: > Sure. This can enforced. I will update the PR, @gziolo.

I didn't realize that the linting rule doesn't allow any other functions to be defined outside of the ones specified in the `prefixes` config property.
Therefore, there's no need to modify the PR to restrict the `gutenberg_` prefixed functions. 🤦‍♂️

The linting rule will only permit functions that start with the `block_core_`, `render_block_core_`, and `register_block_core_ prefixes` (+ the directory name).

The two functions with the `gutenberg_` prefix (`gutenberg_block_core_file_update_interactive_view_script` and `gutenberg_block_core_navigation_update_interactive_view_script`) were allowed because of the `phpcs::disable` statements before them.

Given the discussion above, I'm uncertain about the course of action for these two `gutenberg_` prefixed functions. Should they be allowed for now?

cc @gziolo @luisherranz

File: packages/block-library/src/avatar/index.php
Comment (anton-vlasenko) [reply packages/block-library/src/avatar/index.php]: I've updated the PR.
So it can be reviewed now. 
Thanks!

**Review by gziolo — APPROVED** [review]
I like how the new rule works after a few refinements. In a follow-up we should include guidelines for contributing in `@wordpress/block-library` that explain how functions should be named. There is already a section for that [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/README.md#contributing-to-this-package). We could use that later to reference in the message that the linting rule prints when an error gets detected.

I'm probably not the best person to review the code implementing the rule, so feel free to wait for another review from someone better suited for that task.

**Comment by anton-vlasenko** [comment]
> i'm probably not the best person to review the code implementing the rule, so feel free to wait for another review from someone better suited for that task.

Thank you for the code review, @gziolo.
I understand your concerns. I will ask around for help with code reviews.
Otherwise, I will rebase this PR onto trunk and merge it today evening (UTC+2).
Honestly, I don't see how anything bad could happen if this gets merged.
I've created 2 follow-up issues:
https://github.com/WordPress/gutenberg/issues/53732
https://github.com/WordPress/gutenberg/issues/53731

**Comment by gziolo** [comment]
> I understand your concerns. I will ask around for help with code reviews.
> Otherwise, I will merge this PR today evening (UTC+2).
> Honestly, I don't see how anything bad could happen if this gets merged.

Yes, it should be all good, but I wanted to set proper expectations regarding my review. The plan sounds perfect 😄

**Pull request merged**
